### PR TITLE
Export consumption policies from analytical solution.

### PR DESCRIPTION
**Pull request by MImmesberger**
Export simulated consumption profiles over life-cycles, based on the analytical solution by Iskhakov et al (2017).

### CODE REVIEW

**Comment by timmens** [comment]
The review will be done when issue #6 is solved.

File: tests/test_analytical_simulate.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    # First period
+    {
+        "inputs": {
+            "period": 0,
+            "wealth_levels": np.linspace(1, 100, 12),
+            "wage": 1.0,
+            "interest_rate": 0.0,
+            "work_decision_function": lambda wealth, work_status: True,  # noqa: ARG005
+            "consumption_function": lambda wealth, work_status: wealth,  # noqa: ARG005
+            "work_status_last_period": [True] * 12,
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply tests/test_analytical_simulate.py:22]: `work_decision` is the decision variable, `work_status` the state variable. So, in other words, `work_status` is `work_decision` with a lag.

I could rename `work_status` into `retirement_status`, but then all booleans and how they are processed need to change (`True` to `False` and vice versa). I am not sure whether this is worth the effort, if this is only a minor point?

File: tests/test_analytical_simulate.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        },
+        "expected": {
+            "consumption": np.linspace(1, 100, 12),
+            "work_dec_vec": np.ones(12, dtype=bool),
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply tests/test_analytical_simulate.py]: Fixed.

File: tests/test_analytical_simulate.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    simulate_cons_work_response,
+)
+from numpy.testing import assert_array_almost_equal as aaae
+
+test_cases_simulate_cons_work_resp = [
+    # First period
+    {
+        "inputs": {
+            "period": 0,
+            "wealth_levels": np.linspace(1, 100, 12),
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply tests/test_analytical_simulate.py:16]: Some function expect iterables as an input, some floats. I wanted to explicitly distinguish between those kind of inputs.

File: src/lcm_dev/task_analytical_solution.py [hunk 4]
```diff
@@ -13,40 +13,57 @@
...
         "delta": 1.0,
         "wage": float(100),
-        "r": 0.0,
+        "interest_rate": 0.0,
         "num_periods": 5,
     },
 }
 
 wealth_grid = np.linspace(1, 100, 10_000)
+simulation_grid = np.linspace(1, 100, 20)
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply src/lcm_dev/task_analytical_solution.py]: Fixed.

File: src/lcm_dev/task_analytical_solution.py [hunk 5]
```diff
@@ -13,40 +13,57 @@
         "beta": 0.98,
         "delta": 1.0,
         "wage": float(20),
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply src/lcm_dev/task_analytical_solution.py]: Fixed.

File: tests/test_analytical_simulate.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            "consumption": np.array(
+                [
+                    np.linspace(1, 100, 12),
+                    (np.ones(12, dtype=float) + 1) ** 2,
+                ],
+            ),
+            "work_dec_vec": np.array(
+                [
+                    np.ones(12, dtype=bool),
+                    np.zeros(12, dtype=bool),
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply tests/test_analytical_simulate.py]: Fixed.

File: src/lcm_dev/analytical_solution.py [hunk 7]
```diff
@@ -1,80 +1,288 @@
...
+        wage (float): labor income
+        interest_rate (float): interest rate
+        beta (float): discount factor
+        delta (float): disutility of work
+        tau (int): periods left until end of life
+    Returns:
+        float: retirement threshold
+
+    """
+    k = delta * np.sum([beta**j for j in range(0, tau + 1)]) ** (-1)
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply src/lcm_dev/analytical_solution.py:156]: I don't think that the variable has an economic interpretation. It's the exponent of the function that determines the retirement threshold. I stick to the notation in the Iskhakov et al. (2017).

```python
ret_threshold = ((wage / (1 + interest_rate)) * np.exp(-k)) / (1 - np.exp(-k))
```

File: src/lcm_dev/analytical_solution.py [hunk 8]
```diff
@@ -1,80 +1,288 @@
...
+        consumption_policy (list): consumption policy vector
+        ret_threshold (float): retirement threshold
+        wealth_thresholds (list): wealth thresholds up to `bracket`
+
+    Returns:
+        list: list of wealth thresholds
+
+    """
+    consumption_lb = consumption_policy[bracket + 1]
+    consumption_ub = consumption_policy[bracket + 2]
```
[the remark this answers is not in the document]
Comment (MImmesberger) [reply src/lcm_dev/analytical_solution.py:245]: In this part of the code I compute the wealth level for which the agent is indifferent between functions defined in `consumption_policy`. These indifference points together with the functions defined in `consumption_policy` then define the piecewise consumption function.

The paper yields analytical indifference points for the first threshold of the consumption function. In `_compute_wealth_tresholds` I then do a loop do compute the other thresholds: `for bracket in range(0, (tau - 1) * 2)`. Hence, if `bracket = 0`, I want to compute the threshold implied by consumption function pieces 1 and 2 as I already know the threshold between consumption function piece 0 and 1.

**Comment by MImmesberger** [comment]
Thank you for the extensive review!

I fixed all the issues or left a comment where some more discussion is needed. 

I implemented one more extensive change in [c0b7ba7](https://github.com/OpenSourceEconomics/lcm-dev/pull/5/commits/c0b7ba75650105da4bee385496b6e2da6cc28a52). I thought that it is much more convenient to have several dedicated function for simulation, computation of the value function on a grid and creation of the value and policy functions than to have one big function that returns everything in a dict.

Because I removed a function and changed the content of others, I also adjusted some tests.

**Comment by MImmesberger** [comment]
I also adjusted the test data creation to the new central model defintion in the lcm repo [here](https://github.com/OpenSourceEconomics/lcm-dev/pull/5/commits/bf13cd0b01696265ba51f8b84f5e7d14761ae687).

**Pull request merged**